props/saber_caiwyn_buttons.h

```diff
@@ -0,0 +1,604 @@
+#define PROPS_CAIWYN_BUTTONS_H
+
+#include "prop_base.h"
+#include "../sound/hybrid_font.h"
+#include "../sound/sound_library.h"
+
+#undef PROP_TYPE
+#define PROP_TYPE CaiwynButtons
+
+#ifndef NUM_BUTTONS
+#define NUM_BUTTONS 2
+#endif
+
+#if NUM_BUTTONS < 2
+#error /props/saber_caiwyn_buttons.h requires 2 or 3 Buttons for operation
+#endif
+
+#ifndef BUTTON_DOUBLE_CLICK_TIMEOUT
+#define BUTTON_DOUBLE_CLICK_TIMEOUT 500
+#endif
+
+#ifndef BUTTON_SHORT_CLICK_TIMEOUT
+#define BUTTON_SHORT_CLICK_TIMEOUT 500
+#endif
+
+#ifndef BUTTON_HELD_TIMEOUT
+#define BUTTON_HELD_TIMEOUT 500
+#endif
+
+#ifndef BUTTON_HELD_MEDIUM_TIMEOUT
+#define BUTTON_HELD_MEDIUM_TIMEOUT 1000
+#endif
+
+#ifndef BUTTON_HELD_LONG_TIMEOUT
+#define BUTTON_HELD_LONG_TIMEOUT 2000
+#endif
+
+EFFECT(vmbegin);
+EFFECT(vmend);
+EFFECT(monce);
+EFFECT(mloop);
+EFFECT(mrotate);
+EFFECT(mshuffle);
+
+// The Saber class implements the basic states and actions for the saber.
+class CaiwynButtons : public PropBase {
+public:
+  CaiwynButtons() : PropBase() {}
+  const char* name() override { return "CaiwynButtons"; }
+
+  void Loop() override {
+    PropBase::Loop();
+    DetectMenuTurn();
+    TrackPlayer();
+    sound_library_.Poll(wav_player);
+  }
+
+  virtual void SetPreset(int preset_num, bool announce) override {
+    PropBase::SetPreset(preset_num, announce);
+    strcpy(current_track_,current_preset_.track.get());
+  }
+
+  bool Parse(const char *cmd, const char* arg) override {
+    if (PropBase::Parse(cmd, arg)) return true;
+    if (!strcmp(cmd, "list_current_tracks")) {
+// Tracks must be in: <font>/tracks/*.wav
+      LOCK_SD(true);
+      for (const char* dir = current_directory; dir; dir = next_current_directory(dir)) {
+        PathHelper path(dir, "tracks");
+        ListTracks(path);
+      }
+      LOCK_SD(false);
+      return true;
+    }
+    return false;
+  }
+
+  void VolumeUp() {
+    STDOUT.println("Volume up");
+    if (dynamic_mixer.get_volume() < VOLUME) {
+      dynamic_mixer.set_volume(std::min<int>(VOLUME, dynamic_mixer.get_volume() + VOLUME * 0.10));
+      STDOUT.print("Current Volume: ");
+      STDOUT.println(dynamic_mixer.get_volume());
+    }
+  }
+
+  void VolumeDown() {
+    STDOUT.println("Volume Down");
+    if (dynamic_mixer.get_volume() > (VOLUME * 0.10)) {
+      dynamic_mixer.set_volume(std::max<int>(VOLUME * 0.10, dynamic_mixer.get_volume() - VOLUME * 0.10));
+      STDOUT.print("Current Volume: ");
+      STDOUT.println(dynamic_mixer.get_volume());
+    }
+  }
+
+  void DetectMenuTurn() {
+    if (mode_volume_) {
+      float a = fusor.angle2() - current_menu_angle_;
+      if (a > M_PI) a-=M_PI*2;
+      if (a < -M_PI) a+=M_PI*2;
+      if (a > M_PI / 6) {
+        VolumeUp();
+        current_menu_angle_ = fusor.angle2();
+      }
+      if (a < -M_PI / 6) {
+        VolumeDown();
+        current_menu_angle_ = fusor.angle2();
+      }
+    }
+  }
+
+// Fett263 Track Player, modified by Caiwyn
+  enum TrackMode {
```

Comment:
The instructions at the top say there are three modes, but this shows four?

Reply:
Yes, that's planned future functionality.  I intend to add a shuffle mode eventually, but I haven't figured out how to do it just yet.

Reply:
Removed this for now.